### PR TITLE
New paradigm and layouts and FST!

### LINKED ISSUE #131: Reorganize paradigms into a more intuitive and easy to search order
We will want to shift the plain English and nêhiyawêwin paradigms to the simpler structuring, without the Algonquian division into SAP, Mixed, and Third Person (for Verbs), as user feedback has been that it makes finding the correct form difficult. The Algonquian ordering will  still be accessible in the linguistic *.layout specs.

In practice this means that subjects wlll be presented in Number > Person > Obviation order. And in the case of objects, those will be presented as a cohort, per each subject Number+Person+Obviation feature combination according to the same order.

Something like the following, though I notice that the obviative causes a bit of a headache as to the best natural order.

I -> you (one)
I -> you (all)
I -> him/her
I -> them
I -> him/her/them (further)

You (one) -> me
You (one) -> him/her
You (one) -> us
You (one) -> them
You (one) -> him/her/them (further)

- [x] basic paradigms
- [x] full paradigms

### LINKED ISSUE #392: Revise paradigm and layout files for verbs to match marking of tense as preverb rather than suffix-feature
As of today, the crk FST has been revised so that tense is rather marked with the associated preverb (_kî-_, _wî-_, or _ka-_) rather than as a suffixal feature (`+Prt`, `+Fut+Int`, `+Fut+Def`). This also means that present tense is an unmarked case and  does not have a feature any more (`+Prs`).

This requires that the references to the old tense features anywhere need to be updated to match this new preverb-marking, in the specification of paradigms and layouts as well as in relabeling. In relabeling, we can make use of the standard format for alternative labels of preverbs as well as definitions in CW.

For the verbal paradigms and their layouts, the changes should be deterministic, with the following template:

_ninipân_: `nipâw+V+AI+Ind+Prs+1Sg -> nipâw+V+AI+Ind+1Sg` 
_nikî-nipân_: `nipâw+V+AI+Ind+Prt+1Sg -> PV/ki+nipâw+V+AI+Ind+1Sg` 
_niwî-nipân_: `nipâw+V+AI+Ind+Fut+Int+1Sg -> PV/wi+nipâw+V+AI+Ind+1Sg` 
_nika-nipân_: `nipâw+V+AI+Ind+Fut+Def+1Sg -> PV/ka+nipâw+V+AI+Ind+1Sg `

_ê-nipâyân_: `nipâw+V+AI+Cnj+Prs+1Sg -> PV/e+nipâw+V+AI+Cnj+1Sg` 
_ê-kî-nipâyân_: `nipâw+V+AI+Cnj+Prt+1Sg -> PV/e+PV/ki+nipâw+V+AI+Cnj+1Sg` 
_ê-wî-nipâyân_: `nipâw+V+AI+Cnj+Fut+Int+1Sg -> PV/e+PV/wi+nipâw+V+AI+Cnj+1Sg `

This also means that the affixation of other tense/aspect grammatical preverbs take the unmarked stem case (i.e. what used to be marked as present tense with `+Prs`, but which no longer has that feature/tag). This makes the feature/tag representation of such tense/aspect preverbs more coherent.

### LINKED ISSUE #487: Revise interpretation of *.layout files
Pursuant #392 and #476, the preprocessing of the *.paradigm and *.layout files need to interpret `=` in Conjunct forms so that only the precise, explicitly mentioned preverb (_ê-_, or `PV/e+`) is matched. Otherwise, referring to the unmarked forms, i.e. with only _ê-_ will also match the forms with the tense preverbs, i.e. _ê-kî-_ and _ê-wî-_, which we need to rule out with a strict interpretation of the `=` marker in the layouts.

E.g. `=PV/e+*+V+AI+Cnj+1Sg` should only match `PV/+{{ lemma }}+V+AI+Cnj+1Sg` but
not `PV/+PV/ki+{{ lemma }}+V+AI+Cnj+1Sg` nor `PV/+PV/wi+{{ lemma }}+V+AI+Cnj+1Sg`.

EDIT: removed the `=` sign from the *.paradigm targets (which files we no longer really need, actually).

### LINKED ISSUE #497: Simplify *.paradigm/*.layout files
We've inherited a lot of technical debt from the NDS implementation of \*.layout/\*.paradigm files. Here are some ways we can simplify these files:

 - [x] get rid of the \*.layout files, prefering the \*.layout.csv files instead
 - [x] ~~interpret “smart-quotes” as if they're ASCII quotes~~ (see: #511)
 - [x] get rid of \*.paradigm files, as that can can be inferred from \*.layout.csv
 - [x] changes the extension from \*.layout.csv to \*.layout.tsv because that makes more sense!
---

> In principle and in practice, we wouldn't really want to use the old non-TSV layout files anymore, as all the information should be found in the *.layout.tsv files. On occasion, I've found working with the generated *.layout files useful in that the formatting becomes explicit, and you avoid certain spreadsheet application stupidities, such as the automatic conversion of plain quotes to "smart" (i.e. curly) quotes, if you don't watch out.
> 
> If our itwêwina code, reading the *.layout.tsv files, can interpret alike a plain and a curly quote, then that shouldn't be an issue anymore - and it shouldn't be an issue in the new pane layouts, once we get there.
> 
> Also, do note that the new *.layout.tsv (and *.layout) files do not really need the *.paradigm files, as the `*` (or `=`) can be unequivocally replaced by `{{ lemma }}` (and nothing else).

_Originally posted by @aarppe in https://github.com/UAlbertaALTLab/cree-intelligent-dictionary/pull/496_

### LINKED ISSUE #500: Exclude +Err/Frag analyses for now
Since analyses with `+Err/Frag` are not for "complete" words and may not have a lemma at all, exclude those analyses from the descriptive FST.

### PR DESCRIPTION
**Pull request by eddieantonio**
The new paradigm and layouts for the new FST.

Closes #131 
Closes #392 
Closes #487 
Closes #497 
Closes #500

### CODE REVIEW

**Comment by aarppe** [comment]
N.B. Note that the FSTs need to be updated at the same time, for the paradigm generation to work properly for verbs.

**Comment by aarppe** [comment]
@eddieantonio I'm wondering is this stuck due to some glitch, or would it require some action or fixing, on my or someone else's part? (Note that I didn't move and commit the latest FSTs yet to the ochi-itwêwina project repo).

**Comment by eddieantonio** [comment]
It would require fixing in terms of the coding part. I'll add the latest FST from giella/lang-crk; likely, it will make at least a few tests fail, because of our asumptions on the order of the tags.

**Comment by aarppe** [comment]
The only change should be that tense tags are no longer there (`+Prs`, `+Prt`, `+Fut`, `+Int`, `+Def`), but instead one has more preverb tags (i.e. `PV/ki+`, `PV/wi+`, `PV/ka+`). But as far as I understand, these new grammatical preverb tags should be interpreted exactly like is already done for `PV/e+` etc., and as far as I can imagine, we shouldn't be referring to the old tense tags anywhere. If the latter were still the case, that would seem a potential focus of refactoring.

The bigger change that may require some thought and coding are the fragment analyses, which wouldn't be accepted by the normative generator, and can be filtered based on the `+Err/Frag` (or eventually `+Frag`) tag. There might be an eventual use in getting analyses also for such fragments.

For future reference, some example of fragment analyses (probably need their own issue[s]):

```
ê	PV/e+_+Cnj+Err/Frag	0.000000

ê-kî-nitawi-kâh-kîmôci
ê-kî-nitawi-kâh-kîmôci	PV/e+PV/ki+PV/nitawi+PV/kah+PV/kimoci+_+Cnj+Err/Frag	0.000000
ê-kî-nitawi-kâh-kîmôci	PV/e+PV/ki+PV/nitawi+RdplS+PV/kimoci+_+Cnj+Err/Frag	0.000000

kotiskâwêyâhk
kotiskâwêyâhk	kotiskâwêw+V+AI+Cnj+12Pl+Err/Frag	0.000000
kotiskâwêyâhk	kotiskâwêw+V+AI+Cnj+1Pl+Err/Frag	0.000000
```

**Comment by eddieantonio** [comment]
Actually, I'll want to merge #496 first before continuing on with this, because it will make less of our unit tests fail! 😂

**Comment by eddieantonio** [comment]
> The bigger change that may require some thought and coding are the fragment analyses, which wouldn't be accepted by the normative generator, and can be filtered based on the `+Err/Frag` (or eventually `+Frag`) tag. There might be an eventual use in getting analyses also for such fragments.
> 
> For future reference, some example of fragment analyses (probably need their own issue[s]):
> 
> ```
> ê	PV/e+_+Cnj+Err/Frag	0.000000
> 
> ê-kî-nitawi-kâh-kîmôci
> ê-kî-nitawi-kâh-kîmôci	PV/e+PV/ki+PV/nitawi+PV/kah+PV/kimoci+_+Cnj+Err/Frag	0.000000
> ê-kî-nitawi-kâh-kîmôci	PV/e+PV/ki+PV/nitawi+RdplS+PV/kimoci+_+Cnj+Err/Frag	0.000000
> 
> kotiskâwêyâhk
> kotiskâwêyâhk	kotiskâwêw+V+AI+Cnj+12Pl+Err/Frag	0.000000
> kotiskâwêyâhk	kotiskâwêw+V+AI+Cnj+1Pl+Err/Frag	0.000000
> ```

Haha! This is already causing issues. We have a function called `extract_lemma` that does what it says on the tin. But it gets confused on analyses such as `RdplW+_+Fut+Cond+Err/Frag`, leading to a failure in building the test DB. I'm not terribly sure what the right option is here...

**Comment by eddieantonio** [comment]
Update: here's the output from the database importer:

```
Skipping test word: pe. Could not extract lemma from its analysis: PV/pe+_+Imp+Err/Frag
Skipping test word: pe. Could not extract lemma from its analysis: PV/pe+_+Fut+Cond+Err/Frag
Skipping test word: pe. Could not extract lemma from its analysis: PV/pe+_+Ind+3+Err/Frag
Skipping test word: pe. Could not extract lemma from its analysis: IC+RdplW+_+Cnj+Err/Frag
Skipping test word: ohci. Could not extract lemma from its analysis: PV/ohci+_+Imp+Err/Frag
Skipping test word: ohci. Could not extract lemma from its analysis: PV/ohci+_+Fut+Cond+Err/Frag
Skipping test word: ohci. Could not extract lemma from its analysis: PV/ohci+_+Ind+3+Err/Frag
Skipping test word: sa. Could not extract lemma from its analysis: RdplW+_+Ind+3+Err/Frag
Skipping test word: sa. Could not extract lemma from its analysis: RdplW+_+Imp+Err/Frag
Skipping test word: sa. Could not extract lemma from its analysis: RdplW+_+Fut+Cond+Err/Frag
Skipping test word: poni. Could not extract lemma from its analysis: PV/poni+_+Imp+Err/Frag
Skipping test word: poni. Could not extract lemma from its analysis: PV/poni+_+Fut+Cond+Err/Frag
Skipping test word: poni. Could not extract lemma from its analysis: PV/poni+_+Ind+3+Err/Frag
```

We simply don't have code to deal with `_+Err/Frag`

**Comment by eddieantonio** [comment]
@aarppe: important question:

Is it possible for the FST to output a valid analysis for something that does _not_ have a lemma?

@Madoshakalaka, correct me if I'm wrong, but for items that lack an analysis, we mark it "as-is", and it can be imported into the database with a defintion. However, "as-is" forms never have analyses. If the FST is producing an analysis for things that do not have a lemma, and therefore should be marked "as-is", we have some coding to do!

**Comment by aarppe** [comment]
One way to deal with this is to exclude - for now - _all_ analyses (from the descriptive FST) that contain the feature `+Err/Frag`. Since there is no proper normative for for such fragments.

The reason for having these fragments is that that allows for the scrutiny of multi-part constructions (primarily in syllabics, but in non-standard SRO) so that each part gets its potential analyses that can then be checked for whether they are correctly formed (that the prefixal fragment(s) is/are in potential agreement with the stem fragment).

**Comment by aarppe** [comment]
Suppose someone writes _nikî wâpamâw_ in running text. Of these two segments, _nikî_ and _wâpamâw_ can be analyzed as both independent words but also fragments, cf.

```
niki
niki	nîki+N+I+D+Px1Sg+Sg	0.000000
niki	PV/ki+_+Ind+1+Err/Frag	0.000000

wapamaw
wapamaw	wâpamêw+V+TA+Ind+2Sg+3SgO+Err/Frag	0.000000
wapamaw	wâpamêw+V+TA+Ind+1Sg+3SgO+Err/Frag	0.000000
wapamaw	wâpamêw+V+TA+Ind+X+3SgO+Err/Frag	0.000000
wapamaw	wâpamêw+V+TA+Ind+X+3SgO	0.000000
```

We could then use constraint grammar to choose the maximal construction for which the fragments agree.

```
niki	PV/ki+_+Ind+1+Err/Frag	0.000000

wapamaw	wâpamêw+V+TA+Ind+1Sg+3SgO+Err/Frag	0.000000
```

When using itwêwina instead of running text, we can presume that an input string, even with spaces, is always an entire (lexical) construction. But I suppose with click-in-text, people might be clicking on an individual space-separated string, and we could then provide a partial analysis for that segment - though ideally we'd want to look for the longest non-fragment analysis in the contexts of such a fragment. This can cause challenges in coding, I can imagine.

File: libexec/build_test_db.sh [hunk 1]
```diff
@@ -8,10 +8,12 @@ if [ -f "$DB_FILE" ]; then
   rm -i $DB_FILE
 fi
 
+set -e
+
```
Comment (Madoshakalaka) [inline libexec/build_test_db.sh:12]: This is what I always wanted!! thanks!!

**Comment by eddieantonio** [comment]
Update: after filtering out `+Err/Frag` from some the dictionary importer's FST, things work better. However, for some bizarre reason, the `crk-descriptive-analyzer.hfstol` does not analyze anything. I also tried `analyser-gt-desc.hfstol` from Giella, but the hfstol library is claiming that it's a weighted transducer...? 

I am confusion

**Comment by aarppe** [comment]
Maybe you figured this out, but as far as I know, the FSTs shouldn't be weighted (yet).

**Comment by aarppe** [comment]
@eddieantonio To what extent might it be helpful that a linguist (e.g. I) revised the test cases (now or in the future)? Or is it counterproductive? As in principle every verb analysis/generation will change with this update. (I did revise that Giella YAML tests in April, before committing that FST change over there.)

**Comment by eddieantonio** [comment]
> @eddieantonio To what extent might it be helpful that a linguist (e.g. I) revised the test cases (now or in the future)? Or is it counterproductive? As in principle every verb analysis/generation will change with this update. (I did revise that Giella YAML tests in April, before committing that FST change over there.)

It might be useful if you checked `git grep '[+]Prs` in the repo and figure out which tests/documentation should be updated. I'm not sure that the tests can be updated wholesale though...

**Comment by aarppe** [comment]
There are actually quite a number of places where we still have any of the old tense features:

```
git grep '[+]Prs' | wc -l
     786
git grep '[+]Prt' | wc -l
     303
git grep '[+]Fut[+]Int' | wc -l
     303
git grep '[+]Fut[+]Def' | wc -l
     238
```

In principle and in practice, I can revise the code in those files to match the new FST.

**Comment by aarppe** [comment]
There's also a bunch of references to inflectional categories which are no longer in use, i.e. `VAI-n` and `VAI-v `as well as `VII-n` and `VII-v`.

**Comment by eddieantonio** [comment]
> There are actually quite a number of places where we still have any of the old tense features:
> 
> ```
> git grep '[+]Prs' | wc -l
>      786
> git grep '[+]Prt' | wc -l
>      303
> git grep '[+]Fut[+]Int' | wc -l
>      303
> git grep '[+]Fut[+]Def' | wc -l
>      238
> ```
> 
> In principle and in practice, I can revise the code in those files to match the new FST.

Please ignore any references within `CreeDictionary/tests/utils_tests/**/*` as those are intentionally using the old format. Otherwise, I'll let you see how to deal with these old references.

**Comment by eddieantonio** [comment]
I came up with a nifty `git grep` invocation that cuts things down quite a bit! Check it out:

```sh
 $ git grep -I -F -e +Prs -e +Pst -e +Fut+Int -e +Fut+Def -- ':/!CreeDictionary/tests/utils_tests' | wc -l
      31
```

Only **31** instances! That's enough that we can just list them all!

```sh
 $ PAGER= git grep -I --line-number -F -e +Prs -e +Pst -e +Fut+Int -e +Fut+Def -- ':/!CreeDictionary/tests/utils_tests'
```

```
CreeDictionary/API/models.py:426:                # e.g. Initial change: nêpât: {'IC+nipâw+V+AI+Cnj+Prs+3Sg'}
CreeDictionary/API/models.py:427:                # e.g. Err/Orth: ewapamat: {'PV/e+wâpamêw+V+TA+Cnj+Prs+3Sg+4Sg/PlO+Err/Orth'
CreeDictionary/API/templatetags/inflection_extras.py:21:    >>> presentational_pos({"analysis": "nipâw+V+AI+Ind+Prs+3Sg", "pos": "V", "inflectional_category": "VAI-v"})
CreeDictionary/API/templatetags/inflection_extras.py:23:    >>> presentational_pos({"analysis": "nipâw+V+AI+Ind+Prs+3Sg", "pos": "V", "inflectional_category": ""})
CreeDictionary/API/templatetags/inflection_extras.py:25:    >>> presentational_pos({"analysis": "nipâw+V+AI+Ind+Prs+3Sg", "pos": "", "inflectional_category": "VAI-v"})
CreeDictionary/API/templatetags/inflection_extras.py:27:    >>> presentational_pos({"analysis": "nipâw+V+AI+Ind+Prs+3Sg", "pos": "", "inflectional_category": ""})
CreeDictionary/res/lemma_tags/crk/lemma-tags.tsv:5:VII	${lemma}+V+II+Ind+Prs+3Sg
CreeDictionary/res/lemma_tags/crk/lemma-tags.tsv:6:VAI	${lemma}+V+AI+Ind+Prs+3Sg
CreeDictionary/res/lemma_tags/crk/lemma-tags.tsv:7:VTI	${lemma}+V+TI+Ind+Prs+3Sg
CreeDictionary/res/lemma_tags/crk/lemma-tags.tsv:8:VTA	${lemma}+V+TA+Ind+Prs+3Sg+4Sg/PlO
CreeDictionary/res/lemma_tags/readme.md:5:- VAI lemmas usually have tags +V+AI+Ind+Prs+3Sg
CreeDictionary/tests/DatabaseManager_tests/cmd_test.py:14:        ["yôwamêw+V+TA+Ind+Prs+3Sg+4Sg/PlO"], multi_processing=1, verbose=False
CreeDictionary/tests/DatabaseManager_tests/xml_importer_test.py:14:        ["yôwamêw+V+TA+Ind+Prs+3Sg+4Sg/PlO"], multi_processing=1, verbose=False
CreeDictionary/utils/paradigm_layout_combiner.py:211:                      ['"1s"', 'Ind+Prs+1Sg', 'PV/e+*+Cnj+Prs+1Sg'], ['"2s"', 'Ind+Prs+2Sg', 'PV/e+*+Cnj+Prs+2Sg'],
CreeDictionary/utils/paradigm_layout_combiner.py:212:                      ['"3s"', 'Ind+Prs+3Sg', 'PV/e+*+Cnj+Prs+3Sg'], ['"1p"', 'Ind+Prs+1Pl ', 'PV/e+*+Cnj+Prs+1Pl'],
CreeDictionary/utils/paradigm_layout_combiner.py:213:                      ['"21"', 'Ind+Prs+12Pl', 'PV/e+*+Cnj+Prs+12Pl'], ['"2p"', 'Ind+Prs+2Pl', 'PV/e+*+Cnj+Prs+2Pl'],
CreeDictionary/utils/paradigm_layout_combiner.py:214:                      ['"3p"', 'Ind+Prs+3Pl', 'PV/e+*+Cnj+Prs+3Pl'], ['"4"', 'Ind+Prs+4Sg/Pl', 'PV/e+*+Cnj+Prs+4Sg/Pl'],
CreeDictionary/utils/paradigm_layout_combiner.py:215:                      ['"X"', 'Ind+Prs+X', 'PV/e+*+Cnj+Prs+X'], ['', '', ''], ['', '"PAST TENSE"', ''],
CreeDictionary/utils/paradigm_layout_combiner.py:219:                      ['"1s"', 'Ind+Fut+Int+1Sg', 'PV/e+*+Cnj+Fut+Int+1Sg'],
CreeDictionary/utils/paradigm_layout_combiner.py:220:                      ['"2s"', 'Ind+Fut+Int+2Sg', 'PV/e+*+Cnj+Fut+Int+2Sg'],
CreeDictionary/utils/paradigm_layout_combiner.py:221:                      ['"3s"', 'Ind+Fut+Int+3Sg', 'PV/e+*+Cnj+Fut+Int+3Sg'], ['', '', ''],
CreeDictionary/utils/paradigm_layout_combiner.py:222:                      ['', '"FUTURE DEFINITE TENSE"', ''], ['', ': "Independent"', ''], ['"1s"', 'Ind+Fut+Def+1Sg', ''],
CreeDictionary/utils/paradigm_layout_combiner.py:223:                      ['"2s"', 'Ind+Fut+Def+2Sg', ''], ['"3s"', 'Ind+Fut+Def+3Sg', ''], ['', '', ''],
CreeDictionary/utils/paradigm_layout_combiner.py:229:                   ['"1s"', 'Ind+Prs+1Sg', 'PV/e+*+Cnj+Prs+1Sg'], ['"2s"', 'Ind+Prs+2Sg', 'PV/e+*+Cnj+Prs+2Sg'],
CreeDictionary/utils/paradigm_layout_combiner.py:230:                   ['"3s"', 'Ind+Prs+3Sg', 'PV/e+*+Cnj+Prs+3Sg'], ['"1p"', 'Ind+Prs+1Pl ', 'PV/e+*+Cnj+Prs+1Pl'],
CreeDictionary/utils/paradigm_layout_combiner.py:231:                   ['"21"', 'Ind+Prs+12Pl', 'PV/e+*+Cnj+Prs+12Pl'], ['"2p"', 'Ind+Prs+2Pl', 'PV/e+*+Cnj+Prs+2Pl'],
CreeDictionary/utils/paradigm_layout_combiner.py:232:                   ['"3p"', 'Ind+Prs+3Pl', 'PV/e+*+Cnj+Prs+3Pl'], ['"4"', 'Ind+Prs+4Sg/Pl', 'PV/e+*+Cnj+Prs+4Sg/Pl'],
CreeDictionary/utils/paradigm_layout_combiner.py:233:                   ['"X"', 'Ind+Prs+X', 'PV/e+*+Cnj+Prs+X'], ['', '', ''], ['', '"IMPERATIVE"', ''],
cypress/integration/search.spec.js:129:      // *nipe-acimon == nipê-âcimon == PV/pe+âcimow+V+AI+Ind+Prs+1Sg
docs/glossary.md:23:    saber+V+Pst+Past+1Sg
docs/paradigm-layouts-and-metadata.md:137:| Star substitute   | `*`                 | `PV/e+*+V+TA+Cnj+Prs+1Sg+2SgO` | PV/e+wâpamêw+V+TA+Cnj+Prs+1Sg+2SgO |
```

**Comment by aarppe** [comment]
Ok - I can update and fix the above 31 feature references in the branch, unless instructed otherwise.

**Comment by aarppe** [comment]
We do also have 21 old-fashioned inflectional categories:
```
grep -I -F -e 'VAI-v' -e 'VAI-n' -e 'VII-v' -e 'VII-n' -- ':/!CreeDictionary/tests/utils_tests' | wc -l
      21
```

I'd be inclined to fix those as well (if not now, then soon).

**Comment by aarppe** [comment]
Was able to modify much of the tense stuff, but am not entirely sure how to deal with the code in `CreeDictionary/utils/paradigm_layout_combiner.py`

`CreeDictionary/utils/paradigm_layout_combiner.py:                      ['"1s"', 'Ind+Prs+1Sg', 'PV/e+*+Cnj+Prs+1Sg'], ['"2s"', 'Ind+Prs+2Sg', 'PV/e+*+Cnj+Prs+2Sg'],`

This seems to be to have references to the old *.layout files referring (in an underspecified manner) to the contents of the *.paradigm files. But since we now have completely references, i.e. the *.layout files should spell out exactly what form in the *.paradigm files are referred to, is this actually needed?

**Comment by eddieantonio** [comment]
Blocked by #510

**Comment by eddieantonio** [comment]
Note to Eddie: merging #510 into this should get all the tests to pass. BUT! I'll need to upload a new database manually to Sapir. So do that, Eddie

**Comment by eddieantonio** [comment]
This is finally ready! @aarppe, I'll let you approve, though please do not merge; I have to synchronize this with the database update 😓

**Comment by eddieantonio** [comment]
> The linguistic parts seem to be as they should be, and the code changes seem systematic throughout to a non-coder, though I don't know what they do exactly.

The code changes have been (mostly) approved already in other PRs (e.g., #510)!

**Comment by aarppe** [comment]
**A note to ourselves:** We are using `+Err/Frag` for incomplete components of words, in part because that is what was used as the code in the morphological gold standard, and in part because it simply makes sense that fragments are not accepted or generated as complete word forms. The original Giella scheme rules out any word forms with `+Err/xxx `in their analysis from the "normative" FSTs, which generally appears desirable. However, for spell-checking purposes, we would want to be able to analyze and standardize also fragments, so that these can be reconciled with other fragments to form a complete word (in SRO) or in the case of syllabics, check that the fragments are agreement. Thus, we might eventually want to use the tag `+Frag` instead, since it will allow for the analysis and recognition of fragments by the normative FSTs.

**Pull request merged**